ocs/api_client.py

```diff
@@ -0,0 +1,255 @@
+        lambda x: x.__name__ == client_name,
+        APIClientBase.__subclasses__()
+    )
+
+    try:
+        cls = next(res)
+        return cls()
+    except StopIteration:
+        logger.error(f'Could not find api-client {client_name}')
+        raise
+
+
+class APIClientBase(metaclass=ABCMeta):
+    """
+    Abstract base class for all api-client classes
+
+    This is an abstract base class and api-client specific classes
+    should implement all the methods for interacting with openshift cluster
+    """
+
+    @property
+    @abstractmethod
+    def name(self):
+        """Concrete class will have respective api-client name"""
+        pass
+
+    @abstractmethod
+    def api_get(self):
+        raise NotImplementedError("api_get method is not implemented")
+
+    @abstractmethod
+    def api_post(self):
+        raise NotImplementedError("api_post method is not implemented")
+
+    @abstractmethod
+    def api_delete(self):
+        raise NotImplementedError("api_delete method is not implemented")
+
+    @abstractmethod
+    def api_patch(self):
+        raise NotImplementedError("api_patch method is not implemented")
+
+    @abstractmethod
+    def api_create(self):
+        raise NotImplementedError("api_create method is not implemented")
+
+    @abstractmethod
+    def get_pods(self, **kwargs):
+        """
+        Because of discrepancies in IO format of each client api
+        leaving this to be implemented by specific client
+
+        Args:
+            **kwargs: ex: namespace='namespace',openshift namespace from
+                which we need pods
+
+        Raises:
+            NotImplementedError: if client has not implemented this function.
+
+        Returns:
+            pod_names (list): A list of pod names
+        """
+
+        raise NotImplementedError("get_pods method is not implemented")
+
+    @abstractmethod
+    def get_labels(self, pod_name, pod_namespace):
+        """
+        Get the openshift labels on a given pod
+
+        Args:
+            pod_name(str): Name of pod for which labels to be fetched
+            pod_namespace(str): namespace of pod where this pod lives
+
+        Raises:
+            NotImplementedError: if function not implemented by client
+
+        Returns:
+           dict: Labels associated with pod
+            """
+
+        raise NotImplementedError("get_labels method is not implemented")
+
+    @abstractmethod
+    def create_service(self, **kw):
+        """
+        Create an openshift service
+
+        Args:
+            **kw: ex body="content", namespace='ocnamespace'
+
+        Returns:
+            dict: Response from api server
+
+            Note: Returns could be tricky because it varies from client to
+            client. If its oc-cli then extra work needs to be done in the
+            specific implementation.
+        """
+
+        raise NotImplementedError("get_labels method is not implemented")
+
+
+# All openshift REST client specific imports here
```

Comment:
I would move all imports at top of module and order them in pep8 way.

Reply:
Done